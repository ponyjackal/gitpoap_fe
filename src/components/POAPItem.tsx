import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';

type Props = {
  poap: POAP;
};

export type POAP = {
  created: string;
  owner: string;
  tokenId: string;
  event: {
    city: string;
    country: string;
    description: string;
    end_date: string;
    event_url: string;
    expiry_date: string;
    fancy_id: string;
    id: number;
    image_url: string;
    name: string;
    start_date: string;
    supply: number;
    year: number;
  };
};

const Container = styled.div`
  padding: 10px;
`;

const Image = styled.img`
  max-width: ${rem(180)};
  height: ${rem(180)};
  border-radius: 50%;
`;

export const POAPItem = (props: Props) => {
  const { poap } = props;
  return (
    <Container>
      <Image src={poap.event.image_url} alt={poap.event.description} />
    </Container>
  );
};
