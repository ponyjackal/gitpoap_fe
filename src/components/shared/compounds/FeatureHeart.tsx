import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Loader } from '@mantine/core';
import { FaHeart } from 'react-icons/fa';
import { ExtraHover, ExtraPressed, TextAccent, TextLight } from '../../../colors';
import { useFeaturedPOAPs, useFeaturedPOAPsDispatch } from '../../profile/FeaturedPOAPsContext';

type Props = {
  className?: string;
  poapTokenId: string;
};

const HeartWrapper = styled.div<{ isFeatured: boolean }>`
  > svg {
    fill: ${({ isFeatured }) => (isFeatured ? TextAccent : TextLight)};
  }
`;

const FeatureHeartStyled = styled(FaHeart)`
  color: ${TextLight};
  height: ${rem(18)};
  width: ${rem(18)};
  transition: 150ms fill ease;
  fill: ${TextLight};

  &:hover {
    fill: ${ExtraHover};
    cursor: pointer;
  }
  &:active {
    fill: ${ExtraPressed};
  }
`;

export const FeatureHeart = ({ className, poapTokenId }: Props) => {
  const {
    featuredPOAPsState: { featuredPOAPTokenIDs },
    showHearts,
    loadingIds,
  } = useFeaturedPOAPs();
  const { addFeaturedPOAP, removeFeaturedPOAP } = useFeaturedPOAPsDispatch();
  const isFeatured: boolean = !!featuredPOAPTokenIDs && !!featuredPOAPTokenIDs[poapTokenId];
  const isFeaturedLoading = !!loadingIds && !!loadingIds[poapTokenId];

  if (!showHearts) {
    return null;
  }

  return (
    <HeartWrapper isFeatured={isFeatured}>
      {isFeaturedLoading ? (
        <Loader className={className} size={18} />
      ) : (
        <FeatureHeartStyled
          className={className}
          onClick={() => {
            if (isFeatured) {
              removeFeaturedPOAP(poapTokenId);
            } else {
              addFeaturedPOAP(poapTokenId);
            }
          }}
        />
      )}
    </HeartWrapper>
  );
};
