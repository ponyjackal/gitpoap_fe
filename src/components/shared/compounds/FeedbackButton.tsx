import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Popover, Box } from '@mantine/core';
import { useDisclosure } from '@mantine/hooks';
import { GitPOAPLogoNoText } from '../elements/icons';
import { BackgroundPanel, BackgroundPanel2, ExtraHover, ExtraPressed } from '../../../colors';
import { Text } from '../elements';

const StyledLogo = styled(GitPOAPLogoNoText)`
  height: ${rem(45)};
  width: ${rem(45)};
  transition: transform 200ms ease-in-out;
  path {
    transition: fill 200ms ease-in-out;
  }
`;

const CircleButton = styled.a`
  display: flex;
  align-items: center;
  justify-content: center;
  height: ${rem(70)};
  width: ${rem(70)};
  background-color: ${BackgroundPanel};
  transition: background-color 200ms ease-in-out, transform 200ms ease-in-out;
  border-radius: 50%;
  border: ${rem(3)} solid ${BackgroundPanel2};
  &:hover {
    background-color: ${BackgroundPanel2};
    transform: scale(1.1);
    cursor: pointer;
    ${StyledLogo} {
      transform: scale(1.1);
      path {
        fill: ${ExtraHover};
      }
    }
  }
  &:active {
    background-color: ${BackgroundPanel};
    ${StyledLogo} {
      color: ${ExtraHover};
      path {
        fill: ${ExtraPressed};
      }
    }
  }
`;

const PopoverContainer = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

type Props = {
  className?: string;
  href: string;
};

export const FeedbackButton = ({ className, href }: Props) => {
  const [isOpen, { open, close }] = useDisclosure(false);
  return (
    <Popover
      opened={isOpen}
      onClose={close}
      position="top"
      withArrow
      trapFocus={false}
      closeOnEscape={false}
      transition="fade"
      transitionDuration={200}
      radius="lg"
      withinPortal
      zIndex={100}
      styles={{
        dropdown: {
          backgroundColor: BackgroundPanel,
        },
      }}
    >
      <Popover.Target>
        <Box onMouseEnter={open} onMouseLeave={close} className={className} style={{ zIndex: 100 }}>
          <CircleButton href={href} target="_blank" rel="noopener noreferrer">
            <StyledLogo />
          </CircleButton>
        </Box>
      </Popover.Target>
      <Popover.Dropdown>
        <PopoverContainer>
          <Text>{'Give Feedback'}</Text>
        </PopoverContainer>
      </Popover.Dropdown>
    </Popover>
  );
};
