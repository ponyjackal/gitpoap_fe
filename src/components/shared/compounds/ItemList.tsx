import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Header } from '../elements/Header';
import { Button } from '../elements/Button';
import { FaPlus } from 'react-icons/fa';
import { Select } from '../elements/Select';
import { Text } from '../elements/Text';
import { TextGray } from '../../../colors';
import { Input } from '../elements/Input';

type Props = {
  title?: string;
  className?: string;
  children?: React.ReactNode;
  selectOptions: SelectOption[];
  selectValue: string;
  onSelectChange: (value: string) => void;
  isLoading: boolean;
  hasShowMoreButton: boolean;
  showMoreOnClick: () => void;
  searchInputPlaceholder?: string;
  searchInputValue?: string;
  onSearchInputChange?: React.ChangeEventHandler<HTMLInputElement>;
};

export type SelectOption<T = string> = { value: T; label: string };

const Container = styled.div`
  display: inline-flex;
  flex-direction: column;
  width: 100%;
`;

const Heading = styled.div`
  display: inline-flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
`;

const ListTitle = styled(Header)`
  font-size: ${rem(30)};
  line-height: ${rem(42)};
`;

const Sorting = styled.div`
  display: inline-flex;
  flex-direction: row;
`;

const ShowMore = styled(Button)`
  align-self: center;
`;

const SortBy = styled(Text)`
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: ${rem(12)};
  line-height: ${rem(18)};
  letter-spacing: ${rem(2)};
  text-transform: uppercase;
  color: ${TextGray};
  margin-right: ${rem(10)};
`;

const SearchInput = styled(Input)`
  margin-right: ${rem(50)};
`;

export const ItemList = ({
  children,
  className,
  title,
  selectOptions,
  selectValue,
  onSelectChange,
  isLoading,
  hasShowMoreButton,
  showMoreOnClick,
  searchInputPlaceholder,
  searchInputValue,
  onSearchInputChange,
}: Props) => {
  return (
    <Container className={className}>
      <Heading>
        <ListTitle>{title}</ListTitle>
        <Sorting>
          {searchInputValue !== undefined && onSearchInputChange && (
            <SearchInput
              placeholder={searchInputPlaceholder}
              value={searchInputValue}
              onChange={onSearchInputChange}
            />
          )}
          <SortBy>{'Sort By: '}</SortBy>
          <Select data={selectOptions} value={selectValue} onChange={onSelectChange} />
        </Sorting>
      </Heading>
      {children}
      {hasShowMoreButton && (
        <ShowMore
          onClick={showMoreOnClick}
          leftIcon={<FaPlus />}
          variant="outline"
          loading={isLoading}
        >
          {'Show more'}
        </ShowMore>
      )}
    </Container>
  );
};
