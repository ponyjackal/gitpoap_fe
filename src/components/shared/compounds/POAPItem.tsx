import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import ImageUI from 'next/image';

type Props = {
  poap: POAP;
};

export type POAP = {
  created: string;
  owner: string;
  tokenId: string;
  event: {
    city: string;
    country: string;
    description: string;
    end_date: string;
    event_url: string;
    expiry_date: string;
    fancy_id: string;
    id: number;
    image_url: string;
    name: string;
    start_date: string;
    supply: number;
    year: number;
  };
};

const Container = styled.div`
  padding: ${rem(10)};
`;

const Image = styled(ImageUI)`
  border-radius: 50%;
`;

export const POAPItem = (props: Props) => {
  const { poap } = props;
  return (
    <Container>
      <Image height={180} width={180} src={poap.event.image_url} alt={poap.event.description} />
    </Container>
  );
};
