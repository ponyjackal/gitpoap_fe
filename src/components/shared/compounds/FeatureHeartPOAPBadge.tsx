import React from 'react';
import { Loader } from '../elements/Loader';
import { useFeaturedPOAPsDispatch } from '../../profile/FeaturedPOAPsContext';
import { HeartWrapper, FeatureHeartStyled } from './FeatureHeart';

type Props = {
  className?: string;
  poapTokenId: string;
  isFeatured: boolean;
  isFeaturedLoading: boolean;
};

export const FeaturedHeartPOAPBadge = ({
  className,
  poapTokenId,
  isFeatured,
  isFeaturedLoading,
}: Props) => {
  const { addFeaturedPOAP, removeFeaturedPOAP } = useFeaturedPOAPsDispatch();
  return (
    <HeartWrapper isFeatured={isFeatured}>
      {isFeaturedLoading ? (
        <Loader className={className} size={18} />
      ) : (
        <FeatureHeartStyled
          className={className}
          onClick={() => {
            if (isFeatured) {
              removeFeaturedPOAP(poapTokenId);
            } else {
              addFeaturedPOAP(poapTokenId);
            }
          }}
        />
      )}
    </HeartWrapper>
  );
};
