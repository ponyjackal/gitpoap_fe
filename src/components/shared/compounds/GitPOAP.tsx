import { Link } from '../../Link';
import React from 'react';
import styled, { css } from 'styled-components';
import { rem } from 'polished';
import { GitPOAPBadge } from '../elements/GitPOAPBadge';
import { Title } from '../elements/Title';
import { TextLight } from '../../../colors';
import { FeatureHeart } from './FeatureHeart';

type Props = {
  gitPOAPId: number;
  imgSrc: string;
  name: string;
  orgName: string;
  description?: string;
  className?: string;
  poapTokenId?: string;
  onClick?: () => void;
};

const LineClamp = (lines: number) => css`
  overflow: hidden;
  display: -webkit-box;
  -webkit-line-clamp: ${lines};
  -webkit-box-orient: vertical;
`;

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
`;

const Info = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  max-width: ${rem(150)};
`;

const TitleStyled = styled(Title)`
  margin-top: ${rem(10)};
  text-align: center;
  ${LineClamp(3)};
`;

const OrgName = styled.div`
  font-family: PT Mono;
  font-style: normal;
  font-weight: bold;
  font-size: ${rem(11)};
  line-height: ${rem(18)};
  text-align: center;
  letter-spacing: ${rem(1.2)};
  text-transform: uppercase;
  color: ${TextLight};
  margin-top: ${rem(8)};
`;

const Description = styled.div`
  font-family: PT Mono;
  font-style: normal;
  font-weight: normal;
  font-size: ${rem(11)};
  line-height: ${rem(14)};
  text-align: center;
  letter-spacing: ${rem(-0.1)};
  color: ${TextLight};
  margin-top: ${rem(8)};
  ${LineClamp(2)};
`;

const Heart = styled(FeatureHeart)`
  position: absolute;
  bottom: ${rem(0)};
  right: ${rem(10)};
`;

const BadgeWrapper = styled(Wrapper)`
  position: relative;
`;

export const GitPOAP = ({
  className,
  poapTokenId,
  gitPOAPId,
  imgSrc,
  name,
  orgName,
  description,
  onClick,
}: Props) => {
  return (
    <Wrapper className={className}>
      <BadgeWrapper>
        <Link href={`/gp/${gitPOAPId}`} passHref>
          <GitPOAPBadge
            size="sm"
            imgUrl={imgSrc}
            onClick={() => {
              if (onClick) {
                onClick();
              }
            }}
          />
          {poapTokenId && <Heart poapTokenId={poapTokenId} />}
        </Link>
      </BadgeWrapper>
      <Info>
        <Link href={`/gp/${gitPOAPId}`} passHref>
          <TitleStyled>{name}</TitleStyled>
        </Link>
        <OrgName>{orgName}</OrgName>
        {description && <Description>{description}</Description>}
      </Info>
    </Wrapper>
  );
};
