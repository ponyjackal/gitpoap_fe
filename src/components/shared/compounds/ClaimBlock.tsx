import React from 'react';
import styled from 'styled-components';
import { rem } from 'polished';
import { Button } from '../elements/Button';
import { GitPOAP } from './GitPOAP';
import { FaCheckCircle, FaCoins, FaEthereum } from 'react-icons/fa';

type Props = {
  gitPOAPId: number;
  imgSrc: string;
  name: string;
  orgName: string;
  description: string;
  onClickClaim: () => void;
  onClickBadge?: () => void;
  isClaimed?: boolean;
  isLoading?: boolean;
  isConnected?: boolean;
};

const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
  justify-content: flex-start;
  height: ${rem(340)};
`;

const ButtonWrapper = styled.div`
  flex: 1;
  margin-top: ${rem(14)};
  display: flex;
  flex-direction: column;
  justify-content: flex-end;
`;

const getButtonText = (isClaimed: boolean | undefined, isConnected: boolean | undefined) => {
  if (isClaimed) {
    return 'Minted';
  }

  if (!isConnected) {
    return 'Connect';
  }

  return 'Mint';
};

export const ClaimBlock = ({
  gitPOAPId,
  imgSrc,
  name,
  orgName,
  description,
  onClickClaim,
  onClickBadge,
  isClaimed,
  isLoading,
  isConnected,
}: Props) => {
  return (
    <Wrapper>
      <GitPOAP
        gitPOAPId={gitPOAPId}
        imgSrc={imgSrc}
        name={name}
        orgName={orgName}
        description={description}
        onClick={onClickBadge}
      />
      <ButtonWrapper>
        <Button
          onClick={onClickClaim}
          loading={isLoading}
          leftIcon={isClaimed ? <FaCheckCircle /> : !isConnected ? <FaEthereum /> : <FaCoins />}
          disabled={isClaimed}
        >
          {getButtonText(isClaimed, isConnected)}
        </Button>
      </ButtonWrapper>
    </Wrapper>
  );
};
