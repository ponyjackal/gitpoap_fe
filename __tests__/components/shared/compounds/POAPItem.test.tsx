import { render, screen } from '@testing-library/react';
import 'jest-styled-components';
import { POAPItem, POAP } from '../../../../src/components/shared/compounds/POAPItem';

const renderPOAPItem = (poap: POAP) => {
  return render(<POAPItem poap={poap} />);
};

const testPoap: POAP = {
  created: '09/22/2022',
  owner: 'testOwner',
  tokenId: 'test',
  event: {
    city: '',
    country: '',
    description: 'test-description',
    end_date: '',
    event_url: '',
    expiry_date: '',
    fancy_id: '',
    id: 1,
    image_url: '/test/test.png',
    name: '',
    start_date: '',
    supply: 100,
    year: 2022,
  },
};
describe('POAPItem', () => {
  it('should render poap image', () => {
    renderPOAPItem(testPoap);

    const image = screen.getByRole('img');
    expect(image).toBeInTheDocument();
    expect(image).toHaveAttribute('alt', 'test-description');
    expect(image).toMatchSnapshot();
  });
});
