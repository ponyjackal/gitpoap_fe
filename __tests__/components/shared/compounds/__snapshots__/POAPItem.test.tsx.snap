// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`POAPItem should render poap image 1`] = `
.c0 {
  border-radius: 50%;
}

<img
  alt="test-description"
  class="c0"
  data-nimg="1"
  decoding="async"
  height="180"
  loading="lazy"
  src="/_next/image?url=%2Ftest%2Ftest.png&w=384&q=75"
  srcset="/_next/image?url=%2Ftest%2Ftest.png&w=256&q=75 1x, /_next/image?url=%2Ftest%2Ftest.png&w=384&q=75 2x"
  style="color: transparent;"
  width="180"
/>
`;
